content/r-package-dev/Writing_Tests.Rmd

```diff
@@ -0,0 +1,153 @@
+---
+title: "Writing Tests"
+date: "9999-08-31"
+author: "Jordan I. Walker"
+slug: "writing-tests"
+image: "img/main/intro-icons-300px/r-logo.png"
+output: USGSmarkdowntemplates::hugoTraining
+parent: R Package Development
+weight: 1
+draft: true
+---
+
+```{r setup, include=FALSE, warning=FALSE, message=FALSE}
+library(knitr)
+
+knit_hooks$set(plot=function(x, options) {
+  sprintf("<img src='../%s%s-%d.%s'/ title='%s'/>", 
+          options$fig.path, options$label, options$fig.cur, options$fig.ext, options$fig.cap)
+
+})
+
+opts_chunk$set(
+  echo=TRUE,
+  fig.path="static/writing-tests/",
+  fig.width = 6,
+  fig.height = 6,
+  fig.cap = "TODO"
+)
+
+set.seed(1)
+```
+
+Formal software testing lays down a safety net to ensure that software performs the function it claims to.  Not only will testing your packages improve the confidence level you have in your code, but it will make adding new functionality easier, improve the structure of your code, and save time that would otherwise be used to manually test.  This lesson will go over general principles of testing and how to accomplish them in R using testthat.
+
+## Lesson Objectives
+
+1. Recognize the purpose and value of writing tests
+2. Understand different types of testing and when to apply them
+3. Learn some test writing practices and how often to run them
+
+
+## Why should I test my code?
+
+Formal testing of code is like showing your work on a hard math problem.  You may come to the correct solution, but unless you show exactly how you got to that solution you may not get full credit.  The same is true in software. Adding functionality to a package should be accompanied by code that exercises said functionality and asserts its validity.  Tests are also meant to signal to future developers what the requirements of particular functionality are and to warn when those are not being met.
```

Comment:
I like the Reagan quote :)  I was trying to get to the Dykstra view of software, where it is more like math, but I see that the analogy doesn't quite connect.